### PR TITLE
planner: column pruning must reserve at least one column

### LINKED ISSUE #54777: Strictly checking column pruning would not cause empty column output for each operator
## Enhancement

Zero column after column pruning is quite dangerous. We need to check it during our testing to avoid such cases.

TableDual is a special case. It's bonded with Projection. It only uses virtualRowNum.

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Issue Number: close #54777 

Problem Summary:

### What changed and how does it work?

Column pruning must reserve at least one column for correctness.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/core/logical_projection.go [hunk 1]
```diff
@@ -116,6 +116,23 @@ func (p *LogicalProjection) PruneColumns(parentUsedCols []*expression.Column, op
...
+			break
+		}
+	}
+	if allPruned {
+		_, ok := p.Children()[0].(*LogicalTableDual)
+		if ok {
+			// If the child is dual. The proj should not be eliminated.
+			return p, nil
+		}
+	}
```
Comment (winoros) [inline pkg/planner/core/logical_projection.go:134]: For the `Proj->Dual`.

File: pkg/planner/core/rule_column_pruning.go [hunk 2]
```diff
@@ -155,7 +177,7 @@ func (ds *DataSource) PruneColumns(parentUsedCols []*expression.Column, opt *opt
 	// Current DataSource operator contains all the filters on this table, and the columns used by these filters are always included
 	// in the output schema. Even if they are not needed by DataSource's parent operator. Thus add a projection here to prune useless columns
 	// Limit to MPP tasks, because TiKV can't benefit from this now(projection can't be pushed down to TiKV now).
-	if !addOneHandle && ds.Schema().Len() > len(parentUsedCols) && ds.SCtx().GetSessionVars().IsMPPEnforced() && ds.TableInfo.TiFlashReplica != nil {
+	if !addOneHandle && ds.Schema().Len() > len(parentUsedCols) && len(parentUsedCols) > 0 && ds.SCtx().GetSessionVars().IsMPPEnforced() && ds.TableInfo.TiFlashReplica != nil {
```
Comment (winoros) [inline pkg/planner/core/rule_column_pruning.go:180]: We should not do the project if no column is required from its parent.

File: pkg/planner/core/rule_result_reorder.go [hunk 3]
```diff
@@ -98,7 +101,7 @@ func (rs *resultReorder) injectSort(lp base.LogicalPlan) base.LogicalPlan {
 
 func (*resultReorder) isInputOrderKeeper(lp base.LogicalPlan) bool {
 	switch lp.(type) {
-	case *LogicalSelection, *LogicalProjection, *LogicalLimit:
+	case *LogicalSelection, *LogicalProjection, *LogicalLimit, *LogicalTableDual:
```
Comment (winoros) [inline pkg/planner/core/rule_result_reorder.go:104]: If we don't add this, we'll get `Proj->Sort->Dual` for `Proj->Dual`.

**Review by elsa0520 — APPROVED** [review]
LGTM

File: pkg/planner/core/logical_projection.go (hunk 1 above)
Comment (time-and-fate) [inline pkg/planner/core/logical_projection.go:134]: Please add some comments about why we need this special logic for `TableDual`.

File: pkg/planner/core/logical_join.go [hunk 4]
```diff
@@ -300,13 +300,11 @@ func (p *LogicalJoin) PruneColumns(parentUsedCols []*expression.Column, opt *opt
 	if err != nil {
 		return nil, err
 	}
-	addConstOneForEmptyProjection(p.Children()[0])
```
Comment (time-and-fate) [inline pkg/planner/core/logical_join.go:303]: Now that we remove `addConstOneForEmptyProjection` here, it probably means we can also remove that in `(*LogicalApply).PruneColumns()`. We can have a check.

Comment (winoros) [reply pkg/planner/core/logical_join.go:303]: Yes, i forgot to remove it. Last week is too busy(

**Review by time-and-fate — APPROVED** [review]

**Pull request merged**